### ISSUE TITLE
Error implementing Firestore in Angular v15.

### ISSUE THREAD

**Issue by Nomackleo** [FirebaseExtended/codelab-kanban-fire#26]
Issue:
##

In **Moving the data to Firestore**   I have the following errors::

### Debug Output

-  ERROR TypeError: Cannot read properties of undefined (reading 'valueChanges')
    at new AppComponent (app.component.ts:36:30)
    at NodeInjectorFactory.AppComponent_Factory [as factory] (app.component.ts:24:26)
    at getNodeInjectable (core.mjs:3484:44)
    at createRootComponent (core.mjs:12338:35)
    at ComponentFactory.create (core.mjs:12219:25)
    at ApplicationRef.bootstrap (core.mjs:25315:42)
    at core.mjs:24975:64
    at Array.forEach (<anonymous>)
    at PlatformRef._moduleDoBootstrap (core.mjs:24975:44)
    at core.mjs:24945:26

-  TypeError: Cannot read properties of undefined (reading 'valueChanges')
    at new AppComponent (app.component.ts:36:30)
    at NodeInjectorFactory.AppComponent_Factory [as factory] (app.component.ts:24:26)
    at getNodeInjectable (core.mjs:3484:44)
    at createRootComponent (core.mjs:12338:35)
    at ComponentFactory.create (core.mjs:12219:25)
    at ApplicationRef.bootstrap (core.mjs:25315:42)
    at core.mjs:24975:64
    at Array.forEach (<anonymous>)
    at PlatformRef._moduleDoBootstrap (core.mjs:24975:44)
    at core.mjs:24945:26

### Dependencies en package.json

`"dependencies": {
    "@angular/animations": "^15.1.0",
    "@angular/cdk": "^15.2.0",
    "@angular/common": "^15.1.0",
    "@angular/compiler": "^15.1.0",
    "@angular/core": "^15.1.0",
    "@angular/fire": "^7.5.0",
    "@angular/forms": "^15.1.0",
    "@angular/material": "^15.2.0",
    "@angular/platform-browser": "^15.1.0",
    "@angular/platform-browser-dynamic": "^15.1.0",
    "@angular/router": "^15.1.0",
    "rxjs": "~7.8.0",
    "tslib": "^2.3.0",
    "with": "^7.0.2",
    "zone.js": "~0.12.0"
  },`

### app.component.ts

`import { Component } from '@angular/core';
import { Task } from './task/task';
import { CdkDragDrop, transferArrayItem } from '@angular/cdk/drag-drop';
import { AngularFirestore, AngularFirestoreCollection, AngularFirestoreDocument} from '@angular/fire/compat/firestore';

import { MatDialog } from '@angular/material/dialog';

import { TaskDialogComponent, TaskDialogResult} from './task-dialog/task-dialog.component';
import { Observable } from 'rxjs';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.sass'],
})
export class AppComponent {
  title = 'kanban-fire';
  private todoCollection!: AngularFirestoreCollection<Task>;
  private inProgressCollection!: AngularFirestoreCollection<Task>;
  private doneCollection!: AngularFirestoreCollection<Task>;

  constructor(private dialog: MatDialog, private afs: AngularFirestore) {
    this.todoCollection = this.afs.collection('todo');
    this.inProgressCollection = this.afs.collection('inProgress');
    this.doneCollection = this.afs.collection('done');
  }

  todo = this.todoCollection.valueChanges({
    idField: 'id',
  }) as Observable<Task[]>;

  inProgress = this.inProgressCollection.valueChanges({
    idField: 'id',
  }) as Observable<Task[]>;

  done = this.doneCollection.valueChanges({ idField: 'id' }) as Observable<Task[]>; `

### app.component.html 

`<mat-toolbar color="primary">
  <mat-icon>local_fire_department</mat-icon>
  <span>Kanban Fire</span>
</mat-toolbar>
<div class="content-wrapper">
  <button (click)="newTask()" mat-button>
    <mat-icon>add</mat-icon> Add Task
  </button>
  <div class="container-wrapper">
    <div class="container">
      <h2>Backlog</h2>

      <mat-card
        cdkDropList
        id="todo"
        #todoList="cdkDropList"
        [cdkDropListData]="todo | async"
        [cdkDropListConnectedTo]="[doneList, inProgressList]"
        (cdkDropListDropped)="drop($event)"
        class="list"
      >
        <p class="empty-label" *ngIf="(todo | async)?.length === 0">
          Empty list
        </p>
        <app-task
          (edit)="editTask('todo', $event)"
          *ngFor="let task of todo | async"
          cdkDrag
          [task]="task"
        ></app-task>
      </mat-card>
    </div>

    <div class="container">
      <h2>In progress</h2>

      <mat-card
        cdkDropList
        id="inProgress"
        #inProgressList="cdkDropList"
        [cdkDropListData]="inProgress | async"
        [cdkDropListConnectedTo]="[todoList, doneList]"
        (cdkDropListDropped)="drop($event)"
        class="list"
      >
        <p class="empty-label" *ngIf="(inProgress | async)?.length === 0">
          Empty list
        </p>
        <app-task
          (edit)="editTask('inProgress', $event)"
          *ngFor="let task of inProgress | async"
          cdkDrag
          [task]="task"
        ></app-task>
      </mat-card>
    </div>

    <div class="container">
      <h2>Done</h2>

      <mat-card
        cdkDropList
        id="done"
        #doneList="cdkDropList"
        [cdkDropListData]="done | async"
        [cdkDropListConnectedTo]="[todoList, inProgressList]"
        (cdkDropListDropped)="drop($event)"
        class="list"
      >
        <p class="empty-label" *ngIf="(done | async)?.length === 0">
          Empty list
        </p>
        <app-task
          (edit)="editTask('done', $event)"
          *ngFor="let task of done | async"
          cdkDrag
          [task]="task"
        ></app-task>
      </mat-card>
    </div>
  </div>
</div>`

I really appreciate the help.

**Comment by natereprogle** [comment]
Firestore went through a very large update a while back from v6 to v7. It's recommended you use @angular/fire/compat for the old versions of the API. I'm attempting to migrate this project to v7 and it's giving me so many headaches haha. If I get it figured out I'll let you know. The last piece of the puzzle is converting the app.component.ts lists, but I'm having issues with them not being proper iterables.

**Comment by natereprogle** [comment]
### TL;DR -- [Repo is here](https://github.com/natereprogle/kanban-fire) if you want to see how to create this project using Angular 15 and Angular Fire 7

> Issue:
> 
> In **Moving the data to Firestore** I have the following errors::
> 
> ### Debug Output
> * ERROR TypeError: Cannot read properties of undefined (reading 'valueChanges')
>   at new AppComponent (app.component.ts:36:30)
>   at NodeInjectorFactory.AppComponent_Factory [as factory] (app.component.ts:24:26)
>   at getNodeInjectable (core.mjs:3484:44)
>   at createRootComponent (core.mjs:12338:35)
>   at ComponentFactory.create (core.mjs:12219:25)
>   at ApplicationRef.bootstrap (core.mjs:25315:42)
>   at core.mjs:24975:64
>   at Array.forEach ()
>   at PlatformRef._moduleDoBootstrap (core.mjs:24975:44)
>   at core.mjs:24945:26
> * TypeError: Cannot read properties of undefined (reading 'valueChanges')
>   at new AppComponent (app.component.ts:36:30)
>   at NodeInjectorFactory.AppComponent_Factory [as factory] (app.component.ts:24:26)
>   at getNodeInjectable (core.mjs:3484:44)
>   at createRootComponent (core.mjs:12338:35)
>   at ComponentFactory.create (core.mjs:12219:25)
>   at ApplicationRef.bootstrap (core.mjs:25315:42)
>   at core.mjs:24975:64
>   at Array.forEach ()
>   at PlatformRef._moduleDoBootstrap (core.mjs:24975:44)
>   at core.mjs:24945:26
> 
> ### Dependencies en package.json
> `"dependencies": { "@angular/animations": "^15.1.0", "@angular/cdk": "^15.2.0", "@angular/common": "^15.1.0", "@angular/compiler": "^15.1.0", "@angular/core": "^15.1.0", "@angular/fire": "^7.5.0", "@angular/forms": "^15.1.0", "@angular/material": "^15.2.0", "@angular/platform-browser": "^15.1.0", "@angular/platform-browser-dynamic": "^15.1.0", "@angular/router": "^15.1.0", "rxjs": "~7.8.0", "tslib": "^2.3.0", "with": "^7.0.2", "zone.js": "~0.12.0" },`
> 
> ### app.component.ts
> `import { Component } from '@angular/core'; import { Task } from './task/task'; import { CdkDragDrop, transferArrayItem } from '@angular/cdk/drag-drop'; import { AngularFirestore, AngularFirestoreCollection, AngularFirestoreDocument} from '@angular/fire/compat/firestore';
> 
> import { MatDialog } from '@angular/material/dialog';
> 
> import { TaskDialogComponent, TaskDialogResult} from './task-dialog/task-dialog.component'; import { Observable } from 'rxjs';
> 
> @component({ selector: 'app-root', templateUrl: './app.component.html', styleUrls: ['./app.component.sass'], }) export class AppComponent { title = 'kanban-fire'; private todoCollection!: AngularFirestoreCollection; private inProgressCollection!: AngularFirestoreCollection; private doneCollection!: AngularFirestoreCollection;
> 
> constructor(private dialog: MatDialog, private afs: AngularFirestore) { this.todoCollection = this.afs.collection('todo'); this.inProgressCollection = this.afs.collection('inProgress'); this.doneCollection = this.afs.collection('done'); }
> 
> todo = this.todoCollection.valueChanges({ idField: 'id', }) as Observable<Task[]>;
> 
> inProgress = this.inProgressCollection.valueChanges({ idField: 'id', }) as Observable<Task[]>;
> 
> done = this.doneCollection.valueChanges({ idField: 'id' }) as Observable<Task[]>; `
> 
> ### app.component.html
> ` local_fire_department Kanban Fire
> 
> add Add Task
> ## Backlog
> ```
>   <mat-card
>     cdkDropList
>     id="todo"
>     #todoList="cdkDropList"
>     [cdkDropListData]="todo | async"
>     [cdkDropListConnectedTo]="[doneList, inProgressList]"
>     (cdkDropListDropped)="drop($event)"
>     class="list"
>   >
>     <p class="empty-label" *ngIf="(todo | async)?.length === 0">
>       Empty list
>     </p>
>     <app-task
>       (edit)="editTask('todo', $event)"
>       *ngFor="let task of todo | async"
>       cdkDrag
>       [task]="task"
>     ></app-task>
>   </mat-card>
> </div>
> 
> <div class="container">
>   <h2>In progress</h2>
> 
>   <mat-card
>     cdkDropList
>     id="inProgress"
>     #inProgressList="cdkDropList"
>     [cdkDropListData]="inProgress | async"
>     [cdkDropListConnectedTo]="[todoList, doneList]"
>     (cdkDropListDropped)="drop($event)"
>     class="list"
>   >
>     <p class="empty-label" *ngIf="(inProgress | async)?.length === 0">
>       Empty list
>     </p>
>     <app-task
>       (edit)="editTask('inProgress', $event)"
>       *ngFor="let task of inProgress | async"
>       cdkDrag
>       [task]="task"
>     ></app-task>
>   </mat-card>
> </div>
> 
> <div class="container">
>   <h2>Done</h2>
> 
>   <mat-card
>     cdkDropList
>     id="done"
>     #doneList="cdkDropList"
>     [cdkDropListData]="done | async"
>     [cdkDropListConnectedTo]="[todoList, inProgressList]"
>     (cdkDropListDropped)="drop($event)"
>     class="list"
>   >
>     <p class="empty-label" *ngIf="(done | async)?.length === 0">
>       Empty list
>     </p>
>     <app-task
>       (edit)="editTask('done', $event)"
>       *ngFor="let task of done | async"
>       cdkDrag
>       [task]="task"
>     ></app-task>
>   </mat-card>
> </div>
> ```
> 
> `
> I really appreciate the help.

I've spent about two days figuring this out, and I am happy to say I've got it! I'm going to publish my version of this, and hopefully it'll be of use to somebody. I've used the latest versions of Firebase and Angular as of 3/31/23, so it should be future proof :)

If you want the code now, however, here is the `app.component.ts` file, which should get you were you need to go:

```typescript
import { CdkDragDrop, transferArrayItem } from '@angular/cdk/drag-drop';
import { Component } from '@angular/core';
import {
  Firestore,
  collection,
  doc,
  runTransaction,
  deleteDoc,
  updateDoc,
  addDoc,
  CollectionReference,
  collectionData,
} from '@angular/fire/firestore';
import { MatDialog } from '@angular/material/dialog';
import {
  TaskDialogComponent,
  TaskDialogResult,
} from './task-dialog/task-dialog.component';
import { Task } from './task/task';
import { BehaviorSubject, Observable } from 'rxjs';

const getObservable = (collection: CollectionReference<Task>) => {
  const subject = new BehaviorSubject<Task[]>([]);

  collectionData(collection, { idField: 'id' }).forEach((doc) => {
    subject.next(doc);
  });

  return subject;
};

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css'],
})
export class AppComponent {
  todo = getObservable(
    collection(this.store, 'todo') as CollectionReference<Task>
  ) as Observable<Task[]>;
  inProgress = getObservable(
    collection(this.store, 'inProgress') as CollectionReference<Task>
  ) as Observable<Task[]>;
  done = getObservable(
    collection(this.store, 'done') as CollectionReference<Task>
  ) as Observable<Task[]>;

  editTask(list: 'done' | 'todo' | 'inProgress', task: Task): void {
    const dialogRef = this.dialog.open(TaskDialogComponent, {
      width: '270px',
      data: {
        task,
        enableDelete: true,
      },
    });

    dialogRef
      .afterClosed()
      .subscribe((result: TaskDialogResult | undefined) => {
        if (!result) {
          return;
        }

        if (result.delete) {
          deleteDoc(doc(collection(this.store, list), `${task.id}`));
        } else {
          updateDoc(doc(collection(this.store, list), `${task.id}`), {
            id: task.id,
            title: task.title,
            description: task.description,
          });
        }
      });
  }

  drop(event: CdkDragDrop<Task[] | null, any, any>): void {
    if (event.previousContainer === event.container) {
      return;
    }

    const item = event.previousContainer.data[event.previousIndex] as Task;
    runTransaction(this.store, (transaction) => {
      const promise = Promise.all([
        transaction.delete(
          doc(collection(this.store, event.previousContainer.id), `${item.id}`)
        ),
        transaction.set(
          doc(collection(this.store, event.container.id), `${item.id}`),
          item
        ),
      ]);

      return promise;
    });
    transferArrayItem(
      event.previousContainer.data,
      event.container.data!,
      event.previousIndex,
      event.currentIndex
    );
  }

  constructor(private dialog: MatDialog, private store: Firestore) {}

  newTask(): void {
    const dialogRef = this.dialog.open(TaskDialogComponent, {
      width: '270px',
      data: {
        task: {},
      },
    });

    dialogRef
      .afterClosed()
      .subscribe((result: TaskDialogResult | undefined) => {
        if (!result) {
          return;
        }
        addDoc(collection(this.store, 'todo'), result.task);
      });
  }

  title = 'kanban-fire';
}
```

Be sure you're using the latest versions of the dependencies, and that you are **NOT** using `@angular/fire/compat` _anywhere_.
```json
"dependencies": {
    "@angular/animations": "^15.2.0",
    "@angular/cdk": "15.2.5",
    "@angular/common": "^15.2.0",
    "@angular/compiler": "^15.2.0",
    "@angular/core": "^15.2.0",
    "@angular/fire": "7.5.0",
    "@angular/forms": "^15.2.0",
    "@angular/material": "15.2.5",
    "@angular/platform-browser": "^15.2.0",
    "@angular/platform-browser-dynamic": "^15.2.0",
    "@angular/router": "^15.2.0",
    "rxjs": "~7.8.0",
    "tslib": "^2.3.0",
    "zone.js": "~0.12.0"
  },
```

**Comment by raychenon** [comment]
Ran into the same issues with `Angular: 17.1.2`
@natereprogle  I added a star to your updated [kanban-fire project](https://github.com/natereprogle/kanban-fire)

Thanks

**Comment by natereprogle** [comment]
> Ran into the same issues with `Angular: 17.1.2`
> 
> @natereprogle  I added a star to your updated [kanban-fire project](https://github.com/natereprogle/kanban-fire)
> 
> 
> 
> Thanks

Oh sweet, didn't even realized that worked with Angular 17 haha. Good to know it is 😄
